catalog/admin/module/catalog-admin-module-maplayers/src/main/webapp/map-layers/reducer.js

```diff
@@ -0,0 +1,236 @@
+    headers: {
+      'Content-Type': 'application/json',
+      'Accept': 'application/json'
+    },
+    body: JSON.stringify(body)
+  }
+
+  window.fetch(url, opts)
+    .then((res) => res.text())
+    .then((body) => {
+      const res = JSON.parse(body.replace(/(\[Ljava\.lang\.(Long|String);@[^,]+)/g, (_, str) => `"${str}"`))
+      if (res.status !== 200) {
+        throw new Error(res.error)
+      }
+      dispatch(set(config))
+      dispatch(end())
+      dispatch(message('Successfully saved map layers', 'open intrigue'))
+    })
+    .catch((e) => {
+      dispatch(end())
+      dispatch(message(`Unable to save map layers: ${e.message}`))
+    })
+}
+
+export const validate = (providers) => {
+  let errors = List()
+
+  providers.forEach((layer, i) => {
+    const alpha = layer.get('alpha')
+
+    if (alpha === '') {
+      errors = errors.setIn([i, 'alpha'], 'Alpha cannot be empty')
+    } else if (typeof alpha !== 'number') {
+      errors = errors.setIn([i, 'alpha'], 'Alpha must be a number')
+    } else if (alpha < 0 || alpha > 1) {
+      errors = errors.setIn([i, 'alpha'], 'Invalid alpha')
+    }
+
+    const type = layer.get('type')
+
+    if (type === '') {
+      errors = errors.setIn([i, 'type'], 'Type cannot be empty')
+    } else if (!Object.keys(options).includes(type)) {
+      errors = errors.setIn([i, 'type'], `Invalid type: ${type}`)
+    }
+
+    const url = layer.get('url')
+
+    if (url === '') {
+      errors = errors.setIn([i, 'url'], 'URL cannot be empty')
+    } else if (typeof url !== 'string') {
+      errors = errors.setIn([i, 'url'], 'URL must be a string')
+    } else {
+      const { protocol, host, port } = parse(url)
+
+      if (protocol === null && host === null) {
+        errors = errors.setIn([i, 'url'], 'Invalid URL')
+      } else if (protocol === null) {
+        errors = errors.setIn([i, 'url'], 'Missing protocol')
+      } else if (!protocol.match(/https?:/)) {
+        errors = errors.setIn([i, 'url'], `Unsupported protocol: ${protocol.slice(0, -1)}`)
+      } else if (host === null || host === '') {
+        errors = errors.setIn([i, 'url'], 'URL host cannot be empty')
+      } else if (port < 0 || port > 65535) {
+        errors = errors.setIn([i, 'url'], `Invalid port: ${port}`)
+      }
+    }
+
+    const buffer = layer.get('buffer')
+
+    try {
+      JSON.parse(buffer)
+    } catch (e) {
+      errors = errors.setIn([i, 'buffer'], `Invalid JSON configuration`)
+    }
+  })
+
+  return errors
+}
+
+const emptyLayer = () => fromJS({
+  url: '',
+  type: '',
+  alpha: '',
+  parameters: {}
+})
+
+const providers = (state = List(), { type, path, value = emptyLayer() }) => {
+  switch (type) {
+    case 'map-layers/SET':
+      return value.get('imageryProviders')
+        .map((layer) => layer.set('buffer', JSON.stringify(layer, null, 2)))
+    case 'map-layers/UPDATE':
+      if (value === null) {
+        return state.remove(path[0])
+      } else if (path.includes('buffer')) {
```

Comment:
This update of the imagery provider and buffer looks a bit complex, I think we should leave comments about what is going on. There are also multi values we could extract into a common variable.